### DISCUSSION TITLE
How exactly does running forge script on DeployFundMe deploy an actual FundMe contract to Sepolia?

### DISCUSSION THREAD

**Discussion by dalteko** [Cyfrin/foundry-full-course-cu#186] [Q&A] [answered]
I'm struggling to understand the role of scripts in a general sense. This might not be the best-formatted question because honestly, I don't know what I don't know. 

For example, I can run this code to deploy a fundMe contract on Sepolia 

```
forge script script/DeployFundMe.s.sol:DeployFundMe --rpc-url $(SEPOLIA_RPC_URL) --private-key $(PRIVATE_KEY) --broadcast --verify --etherscan-api-key $(ETHERSCAN_API_KEY) -vvvv
```
But when I dig into the DeployFundMe.s.sol contract, it's simply a contract that returns a fundMe contract with the correct network price feed specification. 

What command is deploying the FundMe contract on Sepolia, and how? It seems weird to me that the command itself doesn't mention the source FundMe.sol contract at all. 

Sorry if this is a stupid question!

----------------

Perhaps what's adding to this confusion is the other scripts: i.e., Interactions.s.sol, which uses the vm.start/stopBroadcast functions, which only work with anvil blockchains, correct? 

Does that mean it's impossible to deploy a contract like FundFundme.sol to Sepolia? 

Sometimes, it feels like scripts are like tests (solely Foundry-based for the most part), while other times they act like a smart contract that one could use on Remix.

Here's the Interactions.s.sol contract (AKA FundFundMe) for reference:

```
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Script, console} from "../lib/forge-std/src/Script.sol";
import {DevOpsTools} from "../lib/foundry-devops/src/DevOpsTools.sol";
import {FundMe} from "../src/FundMe.sol";

contract FundFundMe is Script {
    uint256 constant SEND_VALUE = 0.01 ether;

    function fundFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        vm.stopBroadcast();
    }

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );
        vm.startBroadcast();
        fundFundMe(mostRecentlyDeployed);
        vm.stopBroadcast();
    }
}

contract WithdrawFundMe is Script {
    function withdrawFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).withdraw();
        vm.stopBroadcast();
    }

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );
        vm.startBroadcast();
        withdrawFundMe(mostRecentlyDeployed);
        vm.stopBroadcast();
    }
}
```

**Comment by ghost** [comment, marked as answer]
There are no stupid questions, there are many question so I'll try to give concise anwers.

The objective of scripts is to created pieces of code that execute certains tasks programatically, that's it. What goes inside a script folder? Well usually the frameworks will tell you or recommend you what to put inside them but there is personal preference involved as well but usually you try to create a script to execute a specific task, for example in this course inside the script folder we create the following files:

- Deploy script: this script has the task to programatically deploy a contract.
- Interactions: in this script we create functions to programatically interact with deployed contracts.

But there could be other scripts as well.

The scripts are not meant to be deployed, they are meant to interact with the network by sending transactions using Foundry Cheat Codes like `vm.startBroadcast` and `vm.stopBroadcast.`

**Deploying a contract via a script:**

The following command:

```bash
forge script script/DeployFundMe.s.sol:DeployFundMe --rpc-url $(SEPOLIA_RPC_URL) --private-key $(PRIVATE_KEY) --broadcast --verify --etherscan-api-key $(ETHERSCAN_API_KEY) -vvvv
```

Executes the `DeployFundMe` contract in the `DeployFundMe.s.sol` file inside the script folder, the flags will specify the following:
- `--rpc-url`: on which chain to run the script. In this case on which chain the contract will be deployed.
- `--private-key`: deploying a contract is a transaction and for that you need a private key.
- `--broadcast`: is specific to Foundry, it will send the transacition to the network.
- `--verify`: tells Foundry you want to verify the contract.
- `--etherscan-api-key`: to verify a contract you need to private an api key from the a block explorer associated to the chain.

The command doesn't need to name the `FundMe` contract because when the `DeployFundMe` contract is executed it creates a new contract using the [new](https://docs.soliditylang.org/en/latest/control-structures.html#creating-contracts-via-new) keyword.

```
FundMe fundMe = new FundMe() // creates a new contract
```
`DeployFundMe` can deploy a `FundMe` contract because the source code was imported atthe top:

```bash
import {FundMe} from "../src/FundMe.sol"; // the source code to deploy is known
```

**Comment by pyromaus** [comment]
> What command is deploying the FundMe contract on Sepolia, and how? It seems weird to me that the command itself doesn't mention the source FundMe.sol contract at all.
> 
The command to deploy `FundMe` is in `DeployFundMe`
```solidity
FundMe fundMe = new FundMe(ethUsdPriceFeed);
```
there it is boys. `new <contract>` tells the contract calling it to deploy a new one on whatever chain its on. it seems too simple, I know
> 
> Perhaps what's adding to this confusion is the other scripts: i.e., Interactions.s.sol, which uses the vm.start/stopBroadcast functions, which only work with anvil blockchains, correct?

the `vm` broadcast commands are confusing for sure, you're def not alone there - they are not limited to `anvil` though. Maybe you're thinking of `vm.prank()` - those are anvil-only.

According to Foundry docs, the transactions sandwiched inside `startBroadcast` and `stop` are signed by the person calling the script (you) and not the script itself (since the script is technically a contract deploying the new contract. I hope I got this right)

WITH `broadcast` = running DeployFundMe.sol will make you the deployer
WITHOUT `broadcast` = running DeployFundMe.sol will make DeployFundMe.sol the deployer

> Does that mean it's impossible to deploy a contract like FundFundme.sol to Sepolia?

It is possible. 

Another note, just like with `vm.prank(),` instead of `vm.startPrank(),` `vm.stopPrank()`
You can also use `vm.broadcast()` if you only have one transaction to broadcast. 
That will save a bit of gas in your call
